### DISCUSSION TITLE
Renovate ignores one package in a lerna monorepo

### DISCUSSION THREAD

**Discussion by EricCrosson** [renovatebot/renovate#14912] [Request Help]
Hi! I've been using renovate for a while now and find it invaluable. Y'all have done a great job on this product!

I have one question that's been vexing me for a while, I'm hoping you can shed some light on it.

## What I Did

I'm using the GitHub Renovate App on a private repository, to update dependencies of my npm packages managed in a lerna monorepo. I use `lerna bootstrap` with the --strict[^1] setting, so all packages must declare the same version of a particular dependency or `lerna bootstrap` will fail.

I configured Renovate to run periodically and open pull requests in my repository to update dependency versions across all packages at once.

Here are relevant configuration files from my repository
<details>
  <summary>lerna.json</summary>

  ```json
  {
    "packages": [
      "packages/**",
      "scripts/**"
    ],
    "version": "independent",
    "command": {
      "bootstrap": {
        "hoist": true,
        "forceLocal": true,
        "rejectCycles": true,
        "strict": true
      },
      "link": {
        "forceLocal": true,
        "rejectCycles": true
      }
    }
  }
  ```
</details>

<details>
  <summary>.github/renovate.json</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:js-app"],
  "ignorePresets": [":pinAllExceptPeerDependencies", ":prHourlyLimit2"],
  "labels": ["dependencies"],
  "assignees": ["@ericcrosson"],
  "assignAutomerge": true,
  "prHourlyLimit": 5,
  "hostRules": [
    {
      "timeout": 120000
    }
  ],
  "rangeStrategy": "pin",
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "patch", "pin", "digest"],
      "automerge": false
    },
    {


      "matchDepTypes": ["engines"],
      "rangeStrategy": "auto"
    },
    {
      "groupName": ["aws-cdk packages"],
      "matchPackageNames": ["aws-cdk"],
      "matchPackagePrefixes": ["@aws-cdk/"]
    },
    {
      "groupName": ["unit-testing packages"],
      "matchPackageNames": ["ava", "ava-fast-check", "fast-check"]
    },
    {
      "groupName": ["readable-stream packages"],
      "matchPackageNames": ["@types/readable-stream", "readable-stream"],
      "masterIssueApproval": true,
      "enabled": false
    },
    {
      "matchPackageNames": ["localstack/localstack"],
      "automerge": false
    },
    {
      "matchPackageNames": ["mozjpeg"],
      "enabled": false
    }
  ],
  "timezone": "America/Chicago"
}
```
</details>

## What I Expected to Happen

Each pull request should update the target dependency in _all_ package.json files in the lerna monorepo (so `lerna bootstrap` doesn't fail).

## What Actually Happens

One package (just one!) resists having its dependencies updated, which causes lerna to fail when Renovate tries updating dependencies used in all packages (like eslint or typescript). The name of the package giving resistance is `@stratos-trade/scripts`. 

Here are the debug logs I've been able to collect from the `@stratos-trade/script` package, and a comparison from one other package that does update correctly (any other of the 100+ packages in my monorepo could be chosen to compare against, since they all behave identically)

<details>
  <summary>Debug logs from `@stratos-trade/`</summary>

```json
          {
            "depType": "devDependencies",
            "depName": "typescript",
            "currentValue": "^4.3.5",
            "datasource": "npm",
            "prettyDepType": "devDependency",
            "depIndex": 26,
            "updates": [
              {
                "updateType": "pin",
                "isPin": true,
                "newValue": "4.6.3",
                "newMajor": 4,
                "branchName": "renovate/pin-dependencies"
              }
            ],
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/Microsoft/TypeScript",
            "homepage": "https://www.typescriptlang.org/",
            "currentVersion": "4.6.3"
          }
```
</details>
<details>
  <summary>Debug logs from `@stratos-trade/sqs-data-layer`</summary>

```json
          {
            "depType": "devDependencies",
            "depName": "typescript",
            "currentValue": "^4.3.5",
            "datasource": "npm",
            "prettyDepType": "devDependency",
            "lockedVersion": "4.3.5",
            "depIndex": 29,
            "updates": [
              {
                "updateType": "pin",
                "isPin": true,
                "newValue": "4.3.5",
                "newMajor": 4,
                "branchName": "renovate/pin-dependencies"
              },
              {
                "bucket": "non-major",
                "newVersion": "4.6.3",
                "newValue": "4.6.3",
                "releaseTimestamp": "2022-03-24T22:30:40.329Z",
                "newMajor": 4,
                "newMinor": 6,
                "updateType": "minor",
                "branchName": "renovate/typescript-4.x"
              }
            ],
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/Microsoft/TypeScript",
            "homepage": "https://www.typescriptlang.org/",
            "currentVersion": "4.3.5",
            "isSingleVersion": true,
            "fixedVersion": "4.3.5"
          },
```
</details>

What's interesting is that the `@stratos-trade/scripts` package never calculates updates for the `"bucket": "non-major"` bucket. Looking at my configuration, I can't tell why, but I'd wager I'm overlooking something simple.

For good measure, here are the package.json files for both packages

<details>
  <summary>/scripts/package.json</summary>

```json
{
  "name": "@stratos-trade/scripts",
  "version": "0.0.1",
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "clean": "rimraf --no-glob dist",
    "build": "tsc --build --incremental --verbose .",
    "lint": "eslint --ext json,ts .",
    "lint:fix": "eslint --ext json,ts --fix .",
    "lint-staged": "lint-staged"
  },
  "pre-commit": "lint-staged",
  "dependencies": {
    "@stratos-trade/io-ts": "0.4.0",
    "@stratos-trade/io-ts-lerna": "0.3.0",
    "@typescript-tools/io-ts": "^2.2.0",
    "@typescript-tools/link-local-dependencies": "^1.2.3",
    "@typescript-tools/monorepo-root": "^1.3.1",
    "axios": "^0.21.1",
    "decode-environment": "^1.0.1",
    "execa": "^5.0.0",
    "fp-ts": "^2.10.5",
    "io-ts": "2.1.3",
    "io-ts-docopt": "^1.0.0",
    "ts-pattern": "^2.4.0",
    "yeoman-environment": "^2.10.3"
  },
  "devDependencies": {
    "@ericcrosson/eslint-config": "^2.1.3",
    "@types/yeoman-environment": "^2.10.2",
    "@typescript-eslint/eslint-plugin": "^5.0.0",
    "@typescript-eslint/parser": "^5.0.0",
    "eslint": "^7.28.0",
    "eslint-plugin-json-format": "^2.0.1",
    "lint-staged": "^10.5.3",
    "pre-commit": "^1.2.2",
    "rimraf": "^3.0.2",
    "ts-node": "^9.1.1",
    "typescript": "^4.3.5"
  }
}
```
</details>
<details>
  <summary>/packages/sqs-data-layer/package.json</summary>

```json
{
  "name": "@stratos-trade/sqs-data-layer",
  "version": "0.0.1",
  "private": true,
  "description": "Data layer for SQS operations",
  "homepage": "https://gitlab.com/strong-roots-capital/stratos",
  "bugs": {
    "url": "https://gitlab.com/strong-roots-capital/stratos/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://gitlab.com/strong-roots-capital/stratos.git"
  },
  "license": "UNLICENSED",
  "author": "Eric Crosson <eric.s.crosson@utexas.edu> (https://gitlab.com/eric.s.crosson@utexas.edu)",
  "sideEffects": false,
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "files": [
    "dist/src/"
  ],
  "scripts": {
    "ava": "ava -v",
    "clean": "rimraf --no-glob dist",
    "build": "tsc --build --incremental --verbose .",
    "coverage": "nyc report --all --reporter=text-lcov > coverage.lcov && codecov",
    "docs": "docs-ts",
    "lint": "eslint --ext ts . package.json",
    "lint:fix": "eslint --ext ts --fix . package.json",
    "lint-staged": "lint-staged",
    "test": "nyc ava -v"
  },
  "pre-commit": "lint-staged",
  "dependencies": {
    "@aws-sdk/client-sqs": "^3.2.0",
    "fp-ts": "^2.10.5",
    "io-ts": "2.1.3"
  },
  "devDependencies": {
    "@ericcrosson/eslint-config": "^2.1.3",
    "@types/node": "14.18.12",
    "@types/uuid": "^7.0.2",
    "@typescript-eslint/eslint-plugin": "^5.0.0",
    "@typescript-eslint/parser": "^5.0.0",
    "ava": "3.15.0",
    "ava-fast-check": "4.0.2",
    "codecov": "^3.1.0",
    "docs-ts": "0.6.10",
    "eslint": "^7.28.0",
    "eslint-plugin-ava": "^10.2.0",
    "eslint-plugin-fp-ts": "0.3.0",
    "eslint-plugin-import": "2.25.4",
    "eslint-plugin-json-format": "^2.0.1",
    "eslint-plugin-security": "^1.4.0",
    "fast-check": "2.24.0",
    "lint-staged": "^10.5.3",
    "nyc": "^14.1.0",
    "pre-commit": "^1.2.2",
    "rimraf": "^3.0.2",
    "ts-node": "^9.1.1",
    "typescript": "^4.3.5",
    "uuid": "^7.0.2"
  },
  "ava": {
    "extensions": [
      "ts"
    ],
    "files": [
      "!dist/**/*.d.ts",
      "test/**/test-*.ts"
    ],
    "require": [
      "ts-node/register"
    ],
    "timeout": "2 minutes"
  }
}
```
</details>

And the error will look like this (pasted by Renovate bot into the PR)

<details>
  <summary>Error in PR comment</summary>

```
lerna notice cli v4.0.0
lerna info versioning independent
lerna notice cli v4.0.0
lerna info versioning independent
lerna info Bootstrapping 133 packages
lerna WARN EHOIST_PKG_VERSION "@stratos-trade/scripts" package depends on typescript@^4.3.5, which differs from the hoisted typescript@4.6.3.
lerna ERR! EHOISTSTRICT Package version inconsistencies found while hoisting. Fix the above warnings and retry.
```
</details>

## Postscript

Have I triggered something wonky with the package name, or location of the package? I've read the configuration docs top to bottom several times, and while I've made improvements in Renovate's effectiveness in this monorepo, I've not been able to stop these errors. I also do not have a minimum reproducible repository at this time, I'm hoping the debug logs will spark something for somebody far more experienced than I.

Thanks for reading this and maintaining this project! 

[^1]: https://github.com/lerna/lerna/blob/main/commands/bootstrap/README.md#--strict

**Comment by rarkins** [comment]
The reason for no non-major update is that the `script` dependency has no `lockedVersion` found. So next step I recommend you look into the lock file to see if there's any difference between the two there

**Reply by EricCrosson** [reply]
@rarkins, thanks for the pointer! However, this only raises more questions :sweat_smile: 

You can see in the lerna.json file above that this project uses strict[^1] hoisting[^2], which means the entire repository has just one `package-lock.json` file in the monorepo root, instead of one lockfile per package. And since `--strict` forces the monorepo to use a consistent version of each dependency, that means there is only one declaration of the `typescript` dependency in that single `package-lock.json` file:

```json
    "typescript": {
      "version": "4.3.5",
      "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.3.5.tgz",
      "integrity": "sha512-DqQgihaQ9cUrskJo9kIyW/+g0Vxsk8cDtZ52a3NGh0YNTfpUSArXSohyUGnvbPazEPLu398C0UxmKSOrPumUzA==",
      "dev": true
    },
```

I cannot find any reference to individual package names in the lockfile, either; for example, `@stratos-trade/sqs-data-layer` or `@stratos-trade/scripts`. I am at a loss as to how the `lockedVersion` could be picked up for all packages in the monorepo save one, since it appears this information could only be obtained in one, consistent, place!

[^1]: https://github.com/lerna/lerna/blob/main/commands/bootstrap/README.md#--strict
[^2]: https://github.com/lerna/lerna/blob/main/commands/bootstrap/README.md#--hoist-glob

**Reply by rarkins** [reply]
Unfortunately not much we can do without a public reproduction to debug. Is it the only dependency without a lockedVersion or are there more? Any pattern?

**Reply by EricCrosson** [reply]
The pattern is that sweeping dependency updates leave out modifications to the `@stratos-trade/scripts` package.json in _almost_ every case. For example, I manually added the typescript version diff to the `scripts/package.json` file to get around that error and now we see another

```
lerna notice cli v4.0.0
lerna info versioning independent
lerna info Bootstrapping 133 packages
lerna WARN EHOIST_PKG_VERSION "@stratos-trade/scripts" package depends on fp-ts@^2.10.5, which differs from the hoisted fp-ts@2.11.9.
lerna ERR! EHOISTSTRICT Package version inconsistencies found while hoisting. Fix the above warnings and retry.
```

and another

```
lerna notice cli v4.0.0
lerna info versioning independent
lerna info Bootstrapping 133 packages
lerna WARN EHOIST_PKG_VERSION "@stratos-trade/scripts" package depends on @typescript-eslint/eslint-plugin@^5.0.0, which differs from the hoisted @typescript-eslint/eslint-plugin@5.17.0.
lerna WARN EHOIST_PKG_VERSION "@stratos-trade/scripts" package depends on @typescript-eslint/parser@^5.0.0, which differs from the hoisted @typescript-eslint/parser@5.17.0.
lerna ERR! EHOISTSTRICT Package version inconsistencies found while hoisting. Fix the above warnings and retry.
```

both of these two packages modified 133 files -- the root package.json, and all leaf package.jsons except `scripts/package.json`. In each case that fails to update the `scripts/package.json` file, the debug logs look the same (without the `non-major` section).

Strangely, one PR has exhibited different behavior: upgrading io-ts from "2.1.3" to "2.2.16" _did_ create a diff in the `scripts/package.json` file, and 53 other package.json files besides. I was most surprised by this PR's behavior I observed this morning.

There doesn't seem to be any other monorepo package that sometimes gets left out of the sweeping diffs like `@stratos-trade/scripts` is.

I can see two differences with the first 3 failure cases and the last case:

1. the starting version number was a semver range, and
2. every package required updating

Not sure if those are useful observations though
